implement FormatString for Java 15 String#formatted

codylerum opened the pull request:
Here is an initial pass at an implementation for java 15 and up String.formatted

What I'm doing here is detecting early that this is a formatted call on a string instance and just grabbing the value and all of the arguments and sending them directly to a new validate method.

The old validate method is renamed prepareAndValidate and does the work of extracting the pattern and arguments from the argument this for your traditional String#format calls

fixes #2726

cushon commented on core/src/test/java/com/google/errorprone/bugpatterns/formatstring/FormatStringTest.java:
```diff
@@ -365,4 +365,33 @@ public void invalidIndex() {
+            "  }",
+            "}")
+        .doTest();
+  }
```
Thanks!

cushon commented on core/src/test/java/com/google/errorprone/bugpatterns/formatstring/FormatStringTest.java:
```diff
@@ -16,14 +16,14 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
+import static org.junit.Assume.assumeTrue;
```
nit: can you leave the original import order?

cushon commented on core/src/main/java/com/google/errorprone/bugpatterns/formatstring/FormatString.java:
```diff
@@ -33,17 +36,41 @@
   public Description matchMethodInvocation(MethodInvocationTree tree, final VisitorState state) {
-    ImmutableList<ExpressionTree> args = FormatStringUtils.formatMethodArguments(tree, state);
-    if (args.isEmpty()) {
+    final ImmutableList<ExpressionTree> args;
```
nit: house style is to usually leave `final` off variables, can you leave these out?

(I realize there's a `final VisitorState state` right above this line...)

codylerum commented:
@cushon That worked well and is a much better way of getting the pattern. I'm just not familiar enough with all the tools available when parsing the source code.

I squashed everything into one commit and updated the formatting and import order.
